imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+
+        Parameters
+        ----------
+        range_data : numpy.ndarray
+            Array of range values, where each value is one bit.
+        vectors : numpy.ndarray
+            Array of vectors, where each vector is a tuple of (x, y, z, range).
+            The range value will be overwritten by range_data, and x, y, z will remain
+            the same.
+
+        Returns
+        -------
+        numpy.ndarray
+            Updated array of vectors, identical to vectors with the range values
+            updated from range_data.
+        """
+        if len(range_data) != (len(vectors) - 1) * 2:
+            raise ValueError(
+                "Incorrect length for range_data, there should be two bits per vector, "
+                "excluding the first."
+            )
+
+        updated_vectors: np.ndarray = np.copy(vectors)
+        range_str = "".join([str(i) for i in range_data])
+        for i in range(len(vectors) - 1):
+            range_int = int(range_str[i * 2 : i * 2 + 2], 2)
+            updated_vectors[i + 1][3] = range_int
+        return updated_vectors
+
+    @staticmethod
+    def accumulate_vectors(
+        first_vector: np.ndarray,
+        vector_differences: list[int],
+        vector_count: int,
+    ) -> np.ndarray:
+        """
+        Given a list of differences and the first vector, return calculated vectors.
+
+        This is calculated as follows:
+            vector[i][0] = vector[i-1][0] + vector_differences[i][0]
+            vector[i][1] = vector[i-1][1] + vector_differences[i][1]
+            vector[i][2] = vector[i-1][2] + vector_differences[i][2]
+            vector[i][3] = first_vector[3]
+
+        The third element of the array is the range value, which we assume is the same
+        as the first vector.
+
+        Parameters
+        ----------
+        first_vector : numpy.ndarray
+            A numpy array of 3 signed integers and a range value, representing the
+            start vector.
+        vector_differences : numpy.ndarray
+            A numpy array of shape (expected_vector_count, 4) of signed integers,
+            representing the differences between vectors.
+        vector_count : int
+            The expected number of vectors in the output.
+
+        Returns
+        -------
+        numpy.ndarray
+            A numpy array of shape (expected_vector_count, 4) of signed integers,
+            representing the calculated vectors.
+        """
+        vectors: np.ndarray = np.empty((vector_count, 4), dtype=np.int32)
+        vectors[0] = first_vector
+        if len(vector_differences) % 3 != 0:
+            logger.error(f"Error! Computed compressed vector differences are not "
+                         f"divisible by 3 - meaning some data is missing. "
+                         "Expected length: %s, actual length: "
+                         "%s", vector_count * 3, len(vector_differences))
+        index = 0
+        vector_index = 1
+        for diff in vector_differences:
```

Comment:
Another readability thing but how about in chunks of 3 something like this?
```suggestion
        for vector in [(x,y,z) for x,y,z in np.array(vector_differences).reshape(-1,3)] :
```